x/stake/keeper/validator.go

```diff
@@ -191,132 +191,148 @@ func (k Keeper) ClearTendermintUpdates(ctx sdk.Context) {
+	valPower = GetValidatorsByPowerIndexKey(newValidator, pool)
+	store.Set(valPower, newValidator.Owner)
 
-	// update the validator set for this validator
-	updatedVal := k.UpdateBondedValidators(ctx, validator)
-	if updatedVal.Owner != nil { // updates to validator occurred  to be updated
-		validator = updatedVal
-	}
-	// if decreased in power but still bonded, update Tendermint validator
-	// (if updatedVal is set, the validator has changed bonding status)
-	stillBonded := oldFound && oldValidator.Status() == sdk.Bonded && updatedVal.Owner == nil
-	if stillBonded && oldValidator.PoolShares.Bonded().GT(validator.PoolShares.Bonded()) {
-		bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
-		store.Set(GetTendermintUpdatesKey(ownerAddr), bz)
-	}
-	return validator
+	return valPower
 }
 
-// Update the validator group and kick out any old validators. In addition this
-// function adds (or doesn't add) a validator which has updated its bonded
-// tokens to the validator group. -> this validator is specified through the
-// updatedValidatorAddr term.
+// Update the bonded validator group based on a change to the validator
+// affectedValidator. This function potentially adds the affectedValidator to
+// the bonded validator group which kicks out the cliff validator. Under this
+// situation this function returns the updated affectedValidator.
 //
-// The correct subset is retrieved by iterating through an index of the
-// validators sorted by power, stored using the ValidatorsByPowerIndexKey.
-// Simultaneously the current validator records are updated in store with the
-// ValidatorsBondedIndexKey. This store is used to determine if a validator is a
-// validator without needing to iterate over the subspace as we do in
-// GetValidators.
+// The correct bonded subset of validators is retrieved by iterating through an
+// index of the validators sorted by power, stored using the
+// ValidatorsByPowerIndexKey.  Simultaneously the current validator records are
+// updated in store with the ValidatorsBondedIndexKey. This store is used to
+// determine if a validator is a validator without needing to iterate over all
+// validators.
 //
-// Optionally also return the validator from a retrieve address if the validator has been bonded
 // nolint: gocyclo
 // TODO: Remove the above golint
 func (k Keeper) UpdateBondedValidators(ctx sdk.Context,
-	affectedValidator types.Validator) (updatedVal types.Validator) {
+	affectedValidator types.Validator) (updatedVal types.Validator, updated bool) {
 
 	store := ctx.KVStore(k.storeKey)
 
-	kickCliffValidator := false
 	oldCliffValidatorAddr := k.GetCliffValidator(ctx)
-
-	// add the actual validator power sorted store
 	maxValidators := k.GetParams(ctx).MaxValidators
-	iterator := sdk.KVStoreReversePrefixIterator(store, ValidatorsByPowerIndexKey) // largest to smallest
 	bondedValidatorsCount := 0
-	var validator types.Validator
+	var validator, validatorToBond types.Validator
+	newValidatorBonded := false
```

Comment:
`newValidatorBonded` looks like it could be replaced by `validatorToBond.Owner != nil` (if `validatorToBond` is set)

Reply:
That's how I originally had it, then bez was a bit confused so I figured that it's worth being a bit more verbose and using an extra bool to be more clear. My preference is to leave this

Reply:
OK, either is fine by me